Paper: popmon analysis package for dataset shift detection

sbrugman opened the pull request:
If you are creating this PR in order to submit a draft of your paper,
see http://procbuild.scipy.org/ for logs generated by the build
process.

See the [project readme](https://github.com/scipy-conference/scipy_proceedings#build-your-paper)
for more information.


*Editor*: <!--editor-->@mepa<!--end-editor-->

*Reviewers*: <!--reviewers-list-->@mepa<!--end-reviewers-list-->

sbrugman committed:
initial version paper

scoobies commented:
Hi!, I'm @scoobies, a friendly bot.

Type ```@scoobies help``` to discover how I can help you.

sbrugman committed:
figure(overview): optimize file size for web

sbrugman committed:
title: capitalization

deniederhut commented:
@scoobies mark paper

deniederhut commented:
@scoobies build status

scoobies commented:
{"sbrugman-paper-popmon"=>{"data"=>{"build_output"=>"[*] Updating proceedings build tools in {self.master_repo_path}...\nAlready up to date.\n[*] Check out paper repository...\nCloning into '/tmp/tmp6sexy8zh'...\nMoving proceedings build tools to temp directory...\n[*] Build the paper...\nBuilding: popmon\n*** Warning: /tmp/tmp6sexy8zh/publisher/../output/popmon/paper_stats.json does not exist.\n", "build_pdf_path"=>"/procbuild/procbuild/../cache/sbrugman-paper-popmon.pdf", "build_status"=>"success", "build_timestamp"=>"27/05 17:08"}, "status"=>"success"}}

deniederhut commented:
@scoobies mark needs review

deniederhut commented:
@scoobies assign @mepa as editor

scoobies commented:
Assigned! @mepa is now the editor

deniederhut commented:
Hi! We've moved the deadline for first reviews back one week, to June 16th, to give us a little more time to find reviewers for all the papers this year. You can see the updated full timeline at https://github.com/scipy-conference/scipy_proceedings#timeline-for-2022.

mepa commented:
Hi @sbrugman, we haven't been able to find a reviewer for this paper so I'll be stepping in to review, in the next day or so.

mepa commented:
@scoobies assign @mepa to reviewers

mepa commented:
@scoobies add @mepa to reviewers

scoobies commented:
@mepa added to the reviewers list!

sbrugman commented:
@scoobies check references

scoobies commented:
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1145/2783258.2783359 is OK
- 10.1007/978-0-387-87857-7_2 is OK
- 10.5281/zenodo.61418 is OK
- 10.5281/zenodo.4660177 is OK
- 10.23915/distill.00022 is OK
- 10.1038/s41586-020-2649-2 is OK
- 10.1038/s41592-019-0686-2 is OK
- 10.25080/Majora-92bf1922-00a is OK
- 10.5281/zenodo.3509134 is OK

MISSING DOIs

- 10.4135/9781412961288.n455 may be a valid DOI for title: Teoria statistica delle classi e calcolo delle probabilita
- 10.1007/s10994-018-5719-z may be a valid DOI for title: Reservoir of diverse adaptive learners and stacking fast hoeffding drift detection methods for evolving data streams
- 10.1145/502512.502568 may be a valid DOI for title: A streaming ensemble algorithm (SEA) for large-scale classification
- 10.1145/1014052.1014069 may be a valid DOI for title: Systematic data selection to mine concept-drifting data streams
- 10.5040/9781350207318.00000004 may be a valid DOI for title: Great Expectations
- 10.2307/2333009 may be a valid DOI for title: Continuous inspection schemes

INVALID DOIs

- None
```

sbrugman committed:
chore: include missing DOs + linting

mepa commented on papers/popmon/popmon_paper.rst:
```diff
@@ -0,0 +1,712 @@
+------------
+
+Tracking model performance is crucial to guarantee that a model behaves
+as designed initially. Model performance depends directly on the data
```
"as designed initially" -> Based on the rest of the text in this paragraph, perhaps this could use an additional qualification about the training data, such as "as designed and trained initially"

mepa commented on papers/popmon/popmon_paper.rst:
```diff
@@ -0,0 +1,712 @@
+Tracking model performance is crucial to guarantee that a model behaves
+as designed initially. Model performance depends directly on the data
+used for training and the data predicted on. Changes in the
+latter (e.g. certain word frequency, user demographics, etc.) can affect
```
Without the additional qualification suggested above, it seems that the former (the data used for training) and re-training should also be mentioned. Tracking model performance is also crucial for determining whether to promote a model with the same initial design but trained on different data to production, especially when that decision is made in an automated way.

sbrugman replied:
Agreed, explicitly mentioning the relevance of monitoring model performance with respect to different sets of training data is a welcome addition.

mepa replied:
Thanks.

mepa commented on papers/popmon/popmon_paper.rst:
```diff
@@ -0,0 +1,712 @@
+Related work
+~~~~~~~~~~~~
+
+Many concept drift algorithms exist that follow a similar
```
Since the conference is interdisciplinary and not everybody would know the term "concept drift", I'd recommend defining and/or giving an example of it here or earlier in the text.

sbrugman replied:
Now using the broader term dataset shift, which is explained in the "dataset shift" section.

mepa commented on papers/popmon/popmon_paper.rst:
```diff
@@ -0,0 +1,712 @@
+called ``seldon-deploy``. Integrations for model deployment are
+available based on Kubernetes. The infrastructure setup thus is more
+complex and restrictive than for ``popmon``, which can run on any
+developer's machine.
```
Good job placing popmon in context of other tools.

mepa commented on papers/popmon/popmon_paper.rst:
```diff
@@ -0,0 +1,712 @@
+   target variables (i.e. :math:`p(x | y)`).
+
+Note that there is a lot of variation in terminology used, referring to
+probabilities prevents this ambiguity. For more information on dataset
```
Indeed, this terminology is really nice and clear.

mepa commented on papers/popmon/popmon_paper.rst:
```diff
@@ -0,0 +1,712 @@
+
+``popmon`` is primarily interested in monitoring the distributions of
+features :math:`p(x)` and label :math:`p(y)`, for trained classifiers,
+the data in deployment ideally resembles the training data. However, the package
```
Run-on sentence? Currently there are two subjects: "popmon" and "the data". Please re-work this sentence.

mepa commented on papers/popmon/popmon_paper.rst:
```diff
@@ -0,0 +1,712 @@
+such as kdq-trees :cite:`dasu2006information`. Different
+data representations are in principle compatible with the ``popmon``
+pipeline, as it is similarly structured to alternative methods
+(see :cite:`lu2018learning`, c.f. Fig 5).
```
I don't seem to have access to this reference so can't see Fig. 5. What do "in principle compatible with" and "similarly structured to" mean here? Perhaps an example of using an alternative (to histogram?) data representation with `popmon` could help clarify.

sbrugman replied:
Similarly structured here means having roughly the same steps: data retrieval (reference), data modelling (histogram), test statistics calculation (comparisons), hypothesis testing (static and dynamic bounds). 

For Fig 5, the pre-print is available [here](https://arxiv.org/pdf/2004.05785v1.pdf).

mepa commented on papers/popmon/popmon_paper.rst:
```diff
@@ -0,0 +1,712 @@
+different datasets – the means and correlation between two features are
+identical, but the distributions are different. Histograms of the
+corresponding features (or feature pairs), however, do capture the
+corresponding changes.
```
Nice explanation for the motivation to use histogram-based monitoring instead of summary statistics. What else besides histogram-based monitoring was tried, if anything, but didn't work as well?

mbaak replied:
Initially we went for summary statistics derived directly from the input datasets. But that misses the important comparison of distributions and gave a lot of duplicate code for spark or panda dataframes. We realized that using histograms would solve both problems. Besides we found there are not that many other in-between alternatives available, in particular ones that are supported both by spark and pandas.

mepa replied:
Thanks for the additional context, @mbaak.

mepa commented on papers/popmon/popmon_paper.rst:
```diff
@@ -0,0 +1,712 @@
+Histograms are small in size, and thus are efficiently stored and
+transferred, regardless of the input dataset size. Once data
+records have been aggregated feature-wise, with a minimum number of
+entries per bin, they are typically no longer privacy sensitive.
```
A concrete example may help to illustrate the privacy sensitivity point.

mepa replied:
Thanks for adding the example.

mepa commented on papers/popmon/popmon_paper.rst:
```diff
@@ -0,0 +1,712 @@
+~~~~~~~~~~~~~~
+
+For the creation of histograms from data records the open-source
+``histogrammar`` package has been adopted. ``histogrammar`` has been
```
Were any other packaged considered?

mbaak replied:
Yes, we've looked at multiple. There are some other really good histogramming packages, but none other (that we could find) that support both spark and pandas. In the end we decided to adopt the histogrammar package, since popmon is its main client.

mepa commented on papers/popmon/popmon_paper.rst:
```diff
@@ -0,0 +1,712 @@
+dictionaries with histogram properties. New (index, bin) pairs get
+created whenever needed. Although this could result in out-of-memory
+problems, e.g. when histogramming billions of unique strings, in
+practice this is typically not an issue.
```
Why is this typically not an issue in practice? I imagine it's because a user would not use a feature with billions of unique strings for machine learning. I can also imagine that a user might unintentionally try generating a popmon report, for example, with such a feature, in which case an informative warning/error message could help to diagnose that quickly.

mepa replied:
Thanks, the additional info about how the issue is typically mitigated is really nice.

mepa commented on papers/popmon/popmon_paper.rst:
```diff
@@ -0,0 +1,712 @@
+There is one downside to using histograms: since the data get aggregated
+into bins, and profiles and statistical tests are obtained from the
+histograms, slightly lower resolution is achieved than on the full
+dataset. In practice, however, this is a non-issue; histograms work
```
Are there particular edge cases for which it would be an issue?

sbrugman replied:
We've added:
> The reference type and time-axis binning
 configuration allow the user for selecting an effective resolution.
 
Correctly choosing the binning paramaters resolves this issue in practically all cases.

mepa replied:
Small grammatical fix: 
> The reference type and time-axis binning
> configuration allow the user **to select** an effective resolution.

mepa commented on papers/popmon/popmon_paper.rst:
```diff
@@ -0,0 +1,712 @@
+and comparisons are impossible.
+
+Limitations
+~~~~~~~~~~~
```
Are there any limitations regarding variability in number of records per time range?

sbrugman replied:
The variability of number of records per time range is one of the considerations in which configuration to use, e.g. which reference type is chosen, which comparisons are enabled. It is application dependent whether this should generate an alert.

Some comparison implementations depend on the number of records, other (normalized) comparisons not (apart of perhaps having less statistics).

Once configured, the variability of the number of records per time range is accounted for in the dynamic bounds. When the variability is constant over the reference, then the bounds will be a bit wider.

Note that popmon is quite independent of input (big) dataset size, given that it works both with pandas and spark. After the histogram creation the number of data records per time slot are just counts, which are tiny in comparison.

mepa replied:
Thanks for the clear explanation, @sbrugman.

mepa commented on papers/popmon/popmon_paper.rst:
```diff
@@ -0,0 +1,712 @@
+The comparison can be done in two ways:
+
+#. **Comparisons**: statistically comparing each time slot to the
+   reference data (for example using Kolmogorov-Smirnov, :math:`\chi^2`
```
```suggestion
   reference data (for example using Kolmogorov-Smirnov testing, :math:`\chi^2`
```

mepa commented on papers/popmon/popmon_paper.rst:
```diff
@@ -0,0 +1,712 @@
+data. Moreover, it may pick up an incremental departure (trend) from the initial
+distribution, that will not be significant in comparison to the adjacent
+time-slots. A sliding reference, on the other hand, is updated with more
+recent data, that incorporates this trend. Consider the case where contains a
```
Is this missing something between "where" and "contains"?

mepa commented on papers/popmon/popmon_paper.rst:
```diff
@@ -0,0 +1,712 @@
+~~~~~~~~
+
+Tracking the distribution of values of interest over time is achieved
+via profiles. These are function of the input histogram. Metrics may be
```
```suggestion
via profiles. These are functions of the input histogram. Metrics may be
```

mepa commented on papers/popmon/popmon_paper.rst:
```diff
@@ -0,0 +1,712 @@
+Tracking the distribution of values of interest over time is achieved
+via profiles. These are function of the input histogram. Metrics may be
+defined for all dimensions (e.g. count, correlations), or specifically
+for say 1D numerical histograms (e.g. quantiles). Extending the existing set of
```
Maybe rephrase this as "defined for all dimensions (e.g. count, correlations), or for specific dimensions as in the case of 1D numerical histograms (e.g. quantiles)"?

mepa commented on papers/popmon/popmon_paper.rst:
```diff
@@ -0,0 +1,712 @@
+
+In practice the criteria for normality are hardly ever met, typically
+resulting in a wider distribution with larger tails, yet approximately
+normal behaviour is exhibited. Chebyshev’s
```
Could you please re-work this sentence for grammatical correctness and clarity?

mepa commented on papers/popmon/popmon_paper.rst:
```diff
@@ -0,0 +1,712 @@
+Static monitoring rules
+~~~~~~~~~~~~~~~~~~~~~~~
+
+Static monitoring rules are traditional data quality rules. Denote
```
Add some references to "traditional data quality rules" (e.g., Great Expectations, Deequ)?

sbrugman replied:
A reference to Deequ is now included earlier in the text. "Traditional" here refers to rules that have been in use for decennia. To make this clear, the sentence was extended with an exemplary reference to a relevant paper.

mepa replied:
Thanks!

mepa commented on papers/popmon/popmon_paper.rst:
```diff
@@ -0,0 +1,712 @@
+reports offer multiple views of the data (histograms and heatmaps), the
+profiles and comparisons, and traffic light alerts. There are several
+reasons for providing self-contained reports: they can be opened in the
+browser, easily shared, can be stored as artifacts, and tracked using
```
```suggestion
browser, easily shared, stored as artifacts, and tracked using
```

mepa commented on papers/popmon/popmon_paper.rst:
```diff
@@ -0,0 +1,712 @@
+browser, easily shared, can be stored as artifacts, and tracked using
+tools such as MLFlow. The reports also have no need for an advanced
+infrastructure setup, and are possible to create and view in many
+environments: from a local machine, a (restricted) environment, to public
```
```suggestion
environments: from a local machine, a (restricted) environment, to a public
```

mepa commented on papers/popmon/popmon_paper.rst:
```diff
@@ -0,0 +1,712 @@
+other databases, for example Kibana. One downside of producing self-contained reports is
+that they can get large as the plots are pre-computed and embedded. This
+may be mitigated by replacing the plots with interactive visualizations
+(e.g. using Bokeh or plotly express).
```
References?

sbrugman replied:
Added relevant reference to the text.

mepa commented on papers/popmon/popmon_paper.rst:
```diff
@@ -0,0 +1,712 @@
+artificial datasets: Sine1, Sine2, Mixed, Stagger, Circles, LED, SEA and
+Hyperplane
+:cite:`pesaranghader2018reservoir, street2001streaming, fan2004systematic`.
+These dataset cover myriad dataset shift characteristics: sudden and
```
```suggestion
These datasets cover myriad dataset shift characteristics: sudden and
```

mepa commented on papers/popmon/popmon_paper.rst:
```diff
@@ -0,0 +1,712 @@
+:cite:`pesaranghader2018reservoir, street2001streaming, fan2004systematic`.
+These dataset cover myriad dataset shift characteristics: sudden and
+gradual drifts, order of the relation (dependency of the label on just
+one, or multiple features), binary and multiclass labels and containing
```
```suggestion
one, or multiple features), binary and multiclass labels, and containing
```
(assuming "containing unrelated features" is the fourth characteristic in this list)

mepa commented on papers/popmon/popmon_paper.rst:
```diff
@@ -0,0 +1,712 @@
+
+.. figure:: prev_pearson_led.png
+
+    :label:`pearson1` *LED*: Pearson correlation compared with previous histogram. The shifting points are correctly identified at every 5th of the dataset. Similar patterns are visible for other comparisons, e.g. :math:`\chi^2`.
```
At every 5th _what_ of the dataset?

sbrugman replied:
Clarified this in the text. The statement is about the LED dataset.

mepa commented on papers/popmon/popmon_paper.rst:
```diff
@@ -0,0 +1,712 @@
+
+.. figure:: prev_pearson_sine1.png
+
+    *Sine1*: The dataset shift around data point 20.000, 40.000, 60.000 and 80.000 are clearly visible.
```
```suggestion
    *Sine1*: The dataset shifts around data points 20.000, 40.000, 60.000 and 80.000 are clearly visible.
```

mepa commented on papers/popmon/popmon_paper.rst:
```diff
@@ -0,0 +1,712 @@
+
+    *Sine1*: The dataset shift around data point 20.000, 40.000, 60.000 and 80.000 are clearly visible.
+
+The reports generated by ``popmon`` are capturing features and time bin where
```
```suggestion
The reports generated by ``popmon`` capture features and time bins where
```

mepa commented on papers/popmon/popmon_paper.rst:
```diff
@@ -0,0 +1,712 @@
+feature distributions. The sudden and gradual drifts are clearly visible
+using a rolling reference, see Fig. :ref:`pearson1` for examples.
+The drift in the Hyperplane dataset, incremental and gradual, is not expected to be detected
+using a rolling or self-reference. Moreover, the dataset is synthesized so that the distribution of
```
```suggestion
using a rolling reference or self-reference. Moreover, the dataset is synthesized so that the distribution of
```
(just to be consistent with definitions above)

mepa reviewed (commented):
This paper is really well-organized and pleasant to read. In this initial review I've added suggestions on the very few typos and clarity concerns I found, along with a few questions and comments. My apologies for submitting this late!

deniederhut commented:
Thanks Meghann!

deniederhut commented:
@scoobies mark pending comment

sbrugman committed:
Apply suggestions from code review

Co-authored-by: Meghann Agarwal <agarwal.meghann@gmail.com>

sbrugman committed:
chore: update overview

sbrugman committed:
Feedback processing iteration

sbrugman committed:
chore: reference formatting

sbrugman commented:
@mepa Thank you for your kind words, your time and feedback on the paper. We have answered questions in the threads above, and addressed the feedback in the paper. Happy to continue the discussion if anything is unclear or could be improved.

mepa commented:
@scoobies check references

scoobies commented:
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1109/TKDE.2018.2876857 is OK
- 10.1145/2783258.2783359 is OK
- 10.1007/978-0-387-87857-7_2 is OK
- 10.1007/s10994-018-5719-z is OK
- 10.1145/502512.502568 is OK
- 10.1145/1014052.1014069 is OK
- 10.5281/zenodo.61418 is OK
- 10.5281/zenodo.4660177 is OK
- 10.23915/distill.00022 is OK
- 10.1038/s41586-020-2649-2 is OK
- 10.1038/s41592-019-0686-2 is OK
- 10.1145/2934664 is OK
- 10.25080/Majora-92bf1922-00a is OK
- 10.5281/zenodo.3509134 is OK
- 10.14778/3229863.3229867 is OK

MISSING DOIs

- 10.4135/9781412961288.n455 may be a valid DOI for title: Teoria statistica delle classi e calcolo delle probabilita
- 10.5040/9781350207318.00000004 may be a valid DOI for title: Great Expectations
- 10.2307/2333009 may be a valid DOI for title: Continuous Inspection Schemes

INVALID DOIs

- None
```

mepa commented on papers/popmon/references.bib:
```diff
@@ -0,0 +1,326 @@
+	author       = {Matejka, Justin and Fitzmaurice, George},
+	year         = 2017,
+	booktitle    = {Proceedings of the 2017 CHI conference on human factors in computing systems},
+	pages        = {1290--1294}
```
I think this has a DOI URL of https://doi.org/10.1145/3025453.3025912.

mepa commented on papers/popmon/references.bib:
```diff
@@ -0,0 +1,326 @@
+	author       = {Anscome, F.J.},
+	year         = 1973,
+	journal      = {American Statistician. 27 (1)},
+	pages        = {17--21}
```
https://doi.org/10.2307/2682899

mepa commented on papers/popmon/references.bib:
```diff
@@ -0,0 +1,326 @@
+	author       = {Rabanser, Stephan and G{\"u}nnemann, Stephan and Lipton, Zachary},
+	year         = 2019,
+	journal      = {Advances in Neural Information Processing Systems},
+	volume       = 32
```
@deniederhut , do NeurlIPS (formerly NIPS) papers generally have DOIs? Does https://dl.acm.org/doi/abs/10.5555/3454287.3454412 need to be included in this reference?

mepa replied:
@sbrugman, it looks like this doi is not recognized on doi.org so no need to include it

mepa commented on papers/popmon/references.bib:
```diff
@@ -0,0 +1,326 @@
+	author       = {Gong, Abe and Campbell, James and {Superconductive} and {Great Expectations}},
+	year         = 2022,
+	url          = {https://github.com/great-expectations/great\_expectations},
+	license      = {Apache-2.0}
```
Great Expectations may have a DOI: https://github.com/great-expectations/great_expectations/blob/develop/CITATION.cff

mepa commented on papers/popmon/references.bib:
```diff
@@ -0,0 +1,326 @@
+	pages        = {100--115},
+	issn         = {00063444},
+	url          = {http://www.jstor.org/stable/2333009},
+	urldate      = {2022-06-15}
```
https://doi.org/10.2307/2333009

mepa commented on papers/popmon/popmon_paper.rst:
```diff
@@ -0,0 +1,715 @@
+---------
+
+.. figure:: report_overview.png
+   :scale: 20%
```
Is it possible to increase the size on this figure without it going outside the column width? This might help make the font a bit larger and more readable, particularly when printed on paper.

sbrugman replied:
Removed the scale attribute, now the figure uses the full column width. (Minor rewrite to ensure content fits on page)

mepa commented:
@sbrugman , @mbaak , thank you for a great job addressing my review comments. There are a couple of questions I left unresolved in case you have comments on those that would be relevant to incorporate in the paper. No worries if not. 

Could you please also take another look through to ensure that all citations that have DOIs include them in the .bib file?

And I can confidently recommend this paper for inclusion in the proceedings. Well done!

sbrugman committed:
chore: include additional doi codes

sbrugman commented:
@scoobies check references

scoobies commented:
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1109/TKDE.2018.2876857 is OK
- 10.1145/2783258.2783359 is OK
- 10.1007/978-0-387-87857-7_2 is OK
- 10.1007/s10994-018-5719-z is OK
- 10.1145/502512.502568 is OK
- 10.1145/1014052.1014069 is OK
- 10.5281/zenodo.61418 is OK
- 10.2307/2682899 is OK
- 10.5281/zenodo.4660177 is OK
- 10.23915/distill.00022 is OK
- 10.1038/s41586-020-2649-2 is OK
- 10.1038/s41592-019-0686-2 is OK
- 10.1145/2934664 is OK
- 10.25080/Majora-92bf1922-00a is OK
- 10.5281/zenodo.3509134 is OK
- 10.5281/zenodo.5683574 is OK
- 10.2307/2333009 is OK
- 10.14778/3229863.3229867 is OK

MISSING DOIs

- 10.4135/9781412961288.n455 may be a valid DOI for title: Teoria statistica delle classi e calcolo delle probabilita

INVALID DOIs

- 3025453.3025912 is INVALID
```

sbrugman commented:
@scoobies check references

scoobies commented:
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1109/TKDE.2018.2876857 is OK
- 10.1145/2783258.2783359 is OK
- 10.1007/978-0-387-87857-7_2 is OK
- 10.1007/s10994-018-5719-z is OK
- 10.1145/502512.502568 is OK
- 10.1145/1014052.1014069 is OK
- 10.5281/zenodo.61418 is OK
- 10.2307/2682899 is OK
- 10.1145/3025453.3025912 is OK
- 10.5281/zenodo.4660177 is OK
- 10.23915/distill.00022 is OK
- 10.1038/s41586-020-2649-2 is OK
- 10.1038/s41592-019-0686-2 is OK
- 10.1145/2934664 is OK
- 10.25080/Majora-92bf1922-00a is OK
- 10.5281/zenodo.3509134 is OK
- 10.5281/zenodo.5683574 is OK
- 10.2307/2333009 is OK
- 10.14778/3229863.3229867 is OK

MISSING DOIs

- 10.4135/9781412961288.n455 may be a valid DOI for title: Teoria statistica delle classi e calcolo delle probabilita

INVALID DOIs

- None
```

sbrugman committed:
Increase figure size

Minor rewrite to ensure content fits on page

sbrugman commented:
It was helpful pointing towards the DOIs @mepa, thanks for that. I've incorporated them in the paper. (Apart from the NeurIPS one that we are unsure about.) I also provided a response to the unresolved conversations above.

mepa commented:
@scoobies build paper

scoobies commented:
building paper...

scoobies commented:
{"data"=>{"info"=>"Build for paper sbrugman-paper-popmon scheduled.  Note that builds are only executed if the current build attempt is more than 5 minutes old."}, "status"=>"success"}

mepa commented on papers/popmon/popmon_paper.rst:
```diff
@@ -0,0 +1,714 @@
+environments: from a local machine, a (restricted) environment, to a public
+cloud. If, however, a certain dashboarding tool is available, then the
+metrics computed by ``popmon`` are exposed and can be exported into
+other databases, for example Kibana. One downside of producing self-contained reports is
```
Just noticed a minor thing - should Kibana be referred to as a database?

sbrugman replied:
Included a reference (unfortunately without doi, could not find a relevant code)

mepa replied:
> should Kibana be referred to as a database?

I just realized that this sounds like I'm suggesting to add a reference citing Kibana (which is never a bad thing). What I actually meant to say is that I think of Kibana as a frontend application for Elasticsearch, the database, so was asking about semantics, i.e., "should Kibana be called a database?". No worries either way, and sorry for my poor wording!

mepa commented on papers/popmon/references.bib:
```diff
@@ -0,0 +1,334 @@
+	title        = {Monitoring and explainability of models in production},
+	author       = {Klaise, Janis and Van Looveren, Arnaud and Cox, Clive and Vacanti, Giovanni and Coca, Alexandru},
+	year         = 2020,
+	journal      = {arXiv preprint arXiv:2007.06299}
```
https://doi.org/10.48550/arXiv.2007.06299

mepa commented on papers/popmon/references.bib:
```diff
@@ -0,0 +1,334 @@
+	author       = {Chen, Andrew and Chow, Andy and Davidson, Aaron and DCunha, Arjun and Ghodsi, Ali and Hong, Sue Ann and Konwinski, Andy and Mewald, Clemens and Murching, Siddharth and Nykodym, Tomas and others},
+	year         = 2020,
+	booktitle    = {Proceedings of the fourth international workshop on data management for end-to-end machine learning},
+	pages        = {1--4}
```
https://doi.org/10.1145/3399579.3399867

mepa reviewed (commented):
Hi @sbrugman , just a few more comments from me. I'll go ahead and mark the paper as ready for inclusion in the proceedings, but please do take a look at those.

mepa commented:
@scoobies mark ready

sbrugman committed:
Process feedback

sbrugman commented:
@scoobies check references

scoobies commented:
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1109/TKDE.2018.2876857 is OK
- 10.1145/2783258.2783359 is OK
- 10.1007/978-0-387-87857-7_2 is OK
- 10.1007/s10994-018-5719-z is OK
- 10.1145/502512.502568 is OK
- 10.1145/1014052.1014069 is OK
- 10.5281/zenodo.61418 is OK
- 10.2307/2682899 is OK
- 10.1145/3025453.3025912 is OK
- 10.5281/zenodo.4660177 is OK
- 10.48550/arXiv.2007.06299 is OK
- 10.23915/distill.00022 is OK
- 10.1038/s41586-020-2649-2 is OK
- 10.1038/s41592-019-0686-2 is OK
- 10.1145/2934664 is OK
- 10.25080/Majora-92bf1922-00a is OK
- 10.5281/zenodo.3509134 is OK
- 10.5281/zenodo.5683574 is OK
- 10.1145/3399579.3399867 is OK
- 10.2307/2333009 is OK
- 10.14778/3229863.3229867 is OK

MISSING DOIs

- 10.4135/9781412961288.n455 may be a valid DOI for title: Teoria statistica delle classi e calcolo delle probabilita

INVALID DOIs

- None
```

mepa commented:
Hi @sbrugman and @mbaak! I put in a PR to add popmon to conda-forge during the scipy sprints today so that it can be installed with conda: https://github.com/conda-forge/staged-recipes/pull/19689. Does that sound okay? If you'd like to be listed as maintainer for one or both packages then you can just comment in the PR and I'll add you to the recipe(s).
